packages/browser-pool/src/puppeteer/puppeteer-controller.ts

```diff
@@ -61,7 +63,9 @@ export class PuppeteerController extends BrowserController<
             }
 
             try {
-                const context = await this.browser.createIncognitoBrowserContext(contextOptions);
+                const oldPuppeteerVersion = 'createIncognitoBrowserContext' in CdpBrowser.prototype;
```

Comment:
Couldn't we rely on semver and package.json (which iirc pptr exposes)?

Reply:
We can rely on whatever we want, but reading package.json's versions is often problematic (e.g. not allowed via exports, which I guess is not the case here)

What's wrong with checking for the method? Sounds like a better thing than parsing versions to me. This won't be here forever, we'll drop it with v4.

Reply:
Was just a question mostly, I wouldn't block on this at all. But imo if we need to support multiple versions we should aim to try and read an exposed version if possible before relying on prototype checks

Reply:
Since the prototype check checks exactly the method we use, it actually feels safer to me.

How would you do the version check? Dynamic import into node modules? That sounds much worse, maybe there is a simpler way?

Reply:
Ideally, if a module exposes package.json, we rely on that. Otherwise, if they expose a version constant, we'd rely on that.

Nevertheless this is hopefully not gonna happen too often to require us to do this sort of checks in code :)

Reply:
I don't see any of that here, it's part of the exports field if I read that right, but that's all. So my question stands, how would you do it?

Reply:
I'd probably do something like

```ts
import pptrPkg from 'puppeteer/package.json';

let context: <ContextType>;

if (semver.gte(pptrPkg.version, "22.0.0") {
	context = await browser.<new method>();
} else {
	context = await browser.<old version>();
}
```

but well... that requires casting too, which is bleh. Let's not discuss this more, it doesn't help (plus you said it won't be necessary in v4)

Reply:
I see, but if I try to do that, I get an error about `resolveJsonModule` which we disabled on purpose. I remember disabling that to get around issues with having `dist/src` generated because of importing from root instead of `src` - I can't replicate that right now so maybe things changed, but I would still rather not touch the TS setup right now.